src/darwin/Framework/CHIP/MTRBaseDevice.mm

```diff
@@ -908,8 +936,162 @@ - (void)readAttributesWithEndpointID:(NSNumber * _Nullable)endpointID
+    auto * bridge = new MTRDataValueDictionaryCallbackBridge(queue, completion,
+        ^(ExchangeManager & exchangeManager, const SessionHandle & session, MTRDataValueDictionaryCallback successCb,
+            MTRErrorCallback failureCb, MTRCallbackBridgeBase * bridge) {
+            // interactionStatus tracks whether the whole read interaction has failed.
+            //
+            // Make sure interactionStatus survives even if this block scope is destroyed.
+            auto interactionStatus = std::make_shared<CHIP_ERROR>(CHIP_NO_ERROR);
+
+            auto resultArray = [[NSMutableArray alloc] init];
+            auto onAttributeSuccessCb = [resultArray](const app::ConcreteClusterPath & clusterPath, const AttributeId aAttributeId,
+                                            const MTRDataValueDictionaryDecodableType & aData) {
+                app::ConcreteAttributePath attribPath(clusterPath.mEndpointId, clusterPath.mClusterId, aAttributeId);
+                [resultArray addObject:@ {
+                    MTRAttributePathKey : [[MTRAttributePath alloc] initWithPath:attribPath],
+                    MTRDataKey : aData.GetDecodedObject()
+                }];
+            };
+
+            auto onEventSuccessCb = [resultArray](const app::ConcreteClusterPath & clusterPath, const EventId aEventId,
+                                        const MTRDataValueDictionaryDecodableType & aData) {
+                app::ConcreteEventPath eventPath(clusterPath.mEndpointId, clusterPath.mClusterId, aEventId);
+                [resultArray addObject:@ {
+                    MTREventPathKey : [[MTREventPath alloc] initWithPath:eventPath],
+                    MTRDataKey : aData.GetDecodedObject()
+                }];
+            };
+
+            auto onFailureCb = [resultArray, interactionStatus](
+                                   const app::ConcreteClusterPath * clusterPath, const uint32_t aValueId, CHIP_ERROR aError) {
+                if (clusterPath) {
+                    app::ConcreteAttributePath attribPath(clusterPath->mEndpointId, clusterPath->mClusterId, aValueId);
+                    [resultArray addObject:@ {
+                        MTRAttributePathKey : [[MTRAttributePath alloc] initWithPath:attribPath],
+                        MTRErrorKey : [MTRError errorForCHIPErrorCode:aError]
+                    }];
+                } else {
+                    // This will only happen once per read interaction, and
+                    // after that there will be no more calls to onFailureCb or
+                    // onSuccessCb.
+                    *interactionStatus = aError;
+                }
+            };
+
+            std::vector<app::AttributePathParams> attributePathParamsList;
+            std::vector<app::EventPathParams> eventPathParamsList;
+
+            if (attributePaths != nil) {
+                for (uint8_t i = 0; i < [attributePaths count]; i++) {
+                    chip::EndpointId endpointId = kInvalidEndpointId;
```

Comment:
How about something more like this::

```
AttributePathParams requestPath;
if (attributePath.endpoint != nil) {
  requestPath.endpointId = attributePath.endpoint.unsignedShortValue;
}
if (attributePath.cluster != nil) {
  requestPath.clusterId = static_cast<ClusterId>(attributePath.cluster.unsignedLongValue);
}
if (attributePath.attribute != nil) {
  requestPath.attributeId = static_cast<AttributeId>(attributePath.attribute.unsignedLongValue);
}
attributePathParamsList.push_back(std::move(requestPath));
```
or so.  That way we don't make assumptions about the internal representation of wildcards in AttributePathParams.

Reply:
I modified in AttributeRequestPath, EventRequestPath.